### ISSUE TITLE
Can libtorch be used for quantization-aware training of models?

### ISSUE THREAD

**Issue by hujhcv** [pytorch/pytorch#97985]
### 🚀 The feature, motivation and pitch

We want to use C++ to develop a software for deep learning model training and quantification. Users can fine-tune and train the model by themselves. In order to facilitate deployment, we don't want to use python language, so we want to implement it based on libtorch.
However, libtorch does not seem to support quantization-aware training at present. Is it possible to add this function to libtorch?

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

**Comment by pat749** [comment]
Currently, libtorch supports post-training quantization for CPU and GPU, as well as dynamic quantization for CPU. However, it does not have built-in support for QAT.
To add QAT to libtorch, you will need to implement the quantization-aware training algorithm yourself using the building blocks provided by libtorch, such as the autograd engine, optimizer classes, and data loading utilities. This will likely involve modifying the forward and backward passes of your model to support quantization, as well as implementing quantization-aware versions of the loss function and metrics.

**Comment by hujhcv** [comment]
> 目前，libtorch 支持 CPU 和 GPU 的训练后量化，以及 CPU 的动态量化。但是，它没有对 QAT 的内置支持。要将 QAT 添加到 libtorch，您需要使用 libtorch 提供的构建块（例如 autograd 引擎、优化器类和数据加载实用程序）自己实现量化感知训练算法。这可能涉及修改模型的向前和向后传递以支持量化，以及实现损失函数和指标的量化感知版本。

It seems that implementing QAT by myself is a very complicated work,.  it is better to wait for official support for QAT.

**Comment by supriyar** [comment]
QAT using C++ APIs is currently not supported and we don't have any plans of adding this for now. The official support for QAT in PyTorch can be found in https://pytorch.org/docs/stable/quantization.html#quantization-aware-training-for-static-quantization

**Comment by andrewor14** [comment]
Hi @hujhcv, I'm closing this for now since there are no plans to support this any time in the near future.
